jubilant/_juju.py

```diff
@@ -1057,7 +1057,14 @@ def run(
                     msg = f'timed out waiting for action, stderr:\n{exc.stderr}'
                     raise TimeoutError(msg) from None
                 # The "juju run" CLI command fails if the action has an uncaught exception.
-                if 'task failed' not in exc.stderr:
+                # In Juju 4.0, if the action is not defined, an error like this is returned:
+                #   ERROR adding action operation: adding action operation:
+                #   inserting operation action: inserting action "action-name" for charm
+                #   "app-uuid" and operation "operation-uuid"
+                if (
+                    'task failed' not in exc.stderr
+                    and 'inserting operation action' not in exc.stderr
+                ):
```

Comment:
Juju 3 exits with a zero code, Juju 4 with a non-zero code. I think the Juju 4 behaviour is more correct so Jubilant should handle both.

We could perhaps suggest improvements to the error message that would make it easier for humans to understand and also allow a nicer matching string.

Reply:
Per discussion, let's check that the exception is consistent between Juju 3 and 4. It looks like it'll fall down to the raise ValueError "error running action" for Juju 3 (if return code is zero). But we should check and be consistent.

Reply:
https://github.com/juju/juju/issues/21446